Copy to clipboard button in source block

Vandit1604 opened the pull request:
Fixes #4498.

I have implemented all the things stated in this comment by Dan's [comment](https://github.com/asciidoctor/asciidoctor-extensions-lab/issues/61#issuecomment-1250380863).

However I'm unable to find [Awestruct documentation](http://awestruct.org/) to see how Asciidoctor works under the hood.

krisstern commented:
Hi @Vandit1604 Did you get a chance to check this PR #6183 as a reference?

krisstern commented:
The clipboard button has already been implemented on the Download page, so maybe we could at least use the same icon to consistency? 

<img width="676" alt="Screenshot 2023-04-09 at 2 18 20 PM" src="https://user-images.githubusercontent.com/88480540/230757873-f21fa896-36ae-4c3f-9329-88ec7da06dbd.png">

Vandit1604 commented:
Yes, I checked out #6183 , The Copy to clipboard button on download page was added in a different way since the author of the PR updated the button and added onClick() function. In #4498  [Setting up automated plugin release](https://www.jenkins.io/doc/developer/publishing/releasing-cd/)is generated by .adoc file's content placed in frame.html.haml in _layouts. 

To add copy-to-clipboard button in pages that are generated from .adoc files i Don't think it's possible to use #6183.
In Dan's comment, The steps provided are to add copy-to-clipboard in source code in 'Asciidoctor'.

krisstern commented:
> The clipboard button has already been implemented on the Download page, so maybe we could at least use the same icon for consistency?
> 
> <img alt="Screenshot 2023-04-09 at 2 18 20 PM" width="676" src="https://user-images.githubusercontent.com/88480540/230757873-f21fa896-36ae-4c3f-9329-88ec7da06dbd.png">

@Vandit1604 How about using the same icon used in #6183?

Vandit1604 commented:
Oh, I missed replying about that.

Yeah sure I'll change the icon to match the download page's icon.

krisstern commented on content/css/copy-to-clipboard.css:
```diff
@@ -0,0 +1,91 @@
+.doc .source-toolbox .copy-button.clicked .copy-toast {
+    opacity: 1;
+    transition: none;
+}
```
```suggestion
}

```

NotMyFault reviewed (commented):
Hey,

Thanks for the PR. I'm unsure if it's wise to pick up this item at the current time. This repository is subject to move to Antora during a GSoC project.
Antora contains the added functionality by default, for code blocks and one-liners.
Therefore, it seems a bit redundant to me to move on with it at the moment.

Vandit1604 commented:
> Antora contains the added functionality by default, for code blocks and one-liners. Therefore, it seems a bit redundant to me to move on with it at the moment.

that's what i thought but since the steps were available. I tried implementing it as it is a good to have functionality in jenkins.io.

However Antora provides copy to clipboard out of the box. So should i check out other issues and close this PR (?)

krisstern commented:
I think we could keep this pull request for now, and see if we could achieve the desired effect.

krisstern commented:
@Vandit1604 I have added a `.doc` class to just under the `.col-8` class for the Developer Docs pages via the `_layouts/developer.html.haml` file, but seems like the original source code needs to be tweaked a bit for it to work in our case:

<img width="1414" alt="Screenshot 2023-04-09 at 6 49 02 PM" src="https://user-images.githubusercontent.com/88480540/230768354-94aaa53b-5e23-42d7-a39b-252a31c93fcf.png">

krisstern commented on content/_layouts/frame.html.haml:
```diff
@@ -81,7 +84,8 @@
   %body
     %script{:src => expand_link("/assets/bower/jquery/jquery.min.js")}
-
+    %script{:src => expand_link("/content/js/copy-to-clipboard.js")}
```
```suggestion
    %script{:src => expand_link("/js/copy-to-clipboard.js")}
```

Vandit1604 commented:
@krisstern I tried numerous ways today adding .doc in other places and modifying the css and js. I think i'm missing something. I'll search more(I think on localhost when i execute `make run` browser console shows copy-to-clipboard.js is not found but that's not the case in the deployment of the PR).

I'll push a commit tomorrow if something works or if i make any progress.

krisstern commented:
Hi @Vandit1604 I was able to get the new code to work after modifying the [/js/copy-to-clipboard.js](https://github.com/jenkins-infra/jenkins.io/pull/6260/commits/35ceddf60c950ca2b0983a3326d4e9ee1feb25b0) code at least locally, like so:

<img width="908" alt="Screenshot 2023-04-10 at 4 49 58 PM" src="https://user-images.githubusercontent.com/88480540/230868148-5feea97a-8de7-49d7-9273-15f76ad0892a.png">

krisstern commented:
I have also checked that functionally the button works as expected. But the way I did it was only to enable this feature for the Developer Docs. You may want to add to the other Documentation sections as well if you are feeling ambitious. But at least this should work for for at least the [Setting up automated plugin release](https://www.jenkins.io/doc/developer/publishing/releasing-cd/) page, as #4498 requires. 

I do notice the design of the current SVG icon for copy-to-clipboard may not be ideal, but this is a good start.

krisstern commented:
Just checked that the new feature works in the latest preview too 🎉

Vandit1604 commented:
🎉 This is indeed a nice addition to the site.
I'm out right now. Enabling this feature across the developer docs would be a nice addition. 
When I get back I'll enable it for other pages too.

Vandit1604 commented:
@krisstern I tried to add the copy-to-clipboard functionality to User documentation. It was working but the doc CSS is colliding with page's CSS which makes the page look like this. The sidebar is on the far left and content is on far right with lots of space in middle. 

![image](https://user-images.githubusercontent.com/107131545/230962898-8e64ce2b-c544-4d18-988b-ab6347e842ec.png)

Now I changed the _.doc_ class with _.ctc_ and changed it everywhere but it stopped working on the developer documentation. updating the overall jenkins.css file can break many other things.
Maybe being ambitious won't be possible in this case. 🥲

krisstern commented:
Hi @Vandit1604 No worries, I checked it works after updating to use `.ctc` instead of `.doc` as the special class name for this code change. Have added the feature for the User Guide as well via a new commit. Maybe you just missed updating some class names in the files.

Vandit1604 commented:
If it's working then that's awesome 😎.

> Maybe you just missed updating some class names in the files. 

Yeah that can be the case. Implementation was bit off from my side in this PR.
i apologize for that.

krisstern commented:
@Vandit1604 Are you using VS Code by the way? I like using IntelliJ for programming because at least to me it is more user-friendly out-of-the-box. It helps me catch a lot of bugs especially now that I have GitHub CoPilot enbabled. Its find and replace features are very helpful to me as well.

Vandit1604 commented:
Yes I'm using VS code for jenkins.io PRs. I have never tried Intellij Idea for the same. I'll give it a try.
Thanks for the tip.

krisstern commented:
> Hey,
> 
> Thanks for the PR. I'm unsure if it's wise to pick up this item at the current time. This repository is subject to move to Antora during a GSoC project. Antora contains the added functionality by default, for code blocks and one-liners. Therefore, it seems a bit redundant to me to move on with it currently, given it's subject to major changes within a few months anyway, effectively wiping your changes proposed.

Hi @NotMyFault I can understand your concern totally. But I was thinking if and when the GSoC project goes ahead it will be months away before we can achieve this out-of-the-box experience from Antora. Might as well try and add this feature now since we have just implemented it for the Developer Guide and the User Guide.

krisstern commented:
I think this pull request is ready for a review 🚀

Vandit1604 commented:
@krisstern if there is any more work to be done on this PR please let me know i would love to help.

krisstern commented:
Yeah, agree, I have been thinking about the need to audit each and every User Guide and Developer Guide page to check for inconsistency and also polish up this pull requests to achieve the best effect too.

krisstern commented:
I am confident that this can be accomplished with simple CSS styling changes and simple code changes, such as adding `[source,bash]` or the like where it is missing.

Vandit1604 commented:
Thanks @kmartens27 for taking your time and reviewing it.

>  I was poking around different areas, and it doesn't seem that the copy button is consistent throughout the documentation.
> I also found that there are areas in the Developer documentation that does not provide the copy function: <img alt="no copy button" width="931" src="https://user-images.githubusercontent.com/99040580/231195814-11c06952-f0ce-42f4-9ee3-21f81554dc57.png">

The portions where the copy button is not working we can make changes how @krisstern said by updating the source code or the .adoc pages. The JS uses DOM manipulation to implement copy to clipboard so adding tags like `[source,nameOfLanguage]`

> I found in the regular Jenkins documentation that there was at least one instance where the copy button is overlayed with the chapters/navigation on the right side of the page <img alt="copy overlay" width="1180" src="https://user-images.githubusercontent.com/99040580/231195354-a8b3dd1e-1852-416b-914c-3e704c673ae1.png">

I'll check this i think its something related to CSS i have solved a bug like this previously if that's the case i can be solve it easily.


> And some instances where the copy button/source is overlayed on text within the code block, and it is not clear what is happening. Obviously, the user can figure out what is going on by using other instances on the same page, but on the form validation page for instance, there are 3 different blocks where it is tougher to read <img alt="form validation" width="1371" src="https://user-images.githubusercontent.com/99040580/231196633-e823bff9-dd95-4913-ba35-d421b04d1fbb.png">

The copy button is visible only when you hover over the source block other than that complete command is visible all the time.

> Aside from that, it does not appear that the copy icon has updated with what you've updated, so I was curious as to whether everything was being read correctly.

yes i'm trying to figure that out i have tried multiple svg images but all of them are looking like this. I guess its something to related to CSS. I'm poking around to figure it out.

Vandit1604 commented:
Tomorrow, I'll fix and make all the source code blocks copy-able in User Handbook and other TOC bugs

zbynek commented:
Instead of `<img>` you could create `<ion-icon>` like the other PR does, no need to add a new .svg here. However if implementing/reviewing this requires 

> to audit each and every User Guide and Developer Guide

as @krisstern said, it would be better to abandon this and [wait for the Antora migration](https://github.com/jenkins-infra/jenkins.io/pull/6260#pullrequestreview-1376880124)

Vandit1604 commented:
> Instead of `<img>` you could create `<ion-icon>` like the other PR does, no need to add a new .svg here. However if implementing/reviewing this requires

We have successfully added the copy-to-clipboard on all the pages of Developer Handbook and today i'll do the same for User Handbook.

> > to audit each and every User Guide and Developer Guide

Yeah this is something we can't avoid the feature is implemented all across the User and Developer documentation but due to code block not having the `[source,nameOfLanguage]`. We need to add this in parts where it is not present.

> as @krisstern said, it would be better to abandon this and [wait for the Antora migration](https://github.com/jenkins-infra/jenkins.io/pull/6260#pullrequestreview-1376880124)

Migration to Antora is months away that's why i tried this in the first place. Since the functionality is working without breaking anything in other pages, I think it is a nice to have addition to Jenkins.io.

krisstern commented:
Sounds good!

Vandit1604 commented:
Also another thing to be noted is Antora will only provide the copy to clipboard button to source blocks written like this

```
[source,bash]
$ mvn clean verify
```

without the `[source,bash]` it won't add the copy button to it.

Vandit1604 commented:
TOC Problem is in pages

- https://www.jenkins.io/doc/book/installing/other/#other-operating-systems
- https://www.jenkins.io/doc/book/pipeline/jenkinsfile/
- https://www.jenkins.io/doc/book/pipeline/development/
- https://www.jenkins.io/doc/book/pipeline/syntax/

reducing the size of source code block can solve the problem on all the pages.

Reverse Proxy configuration page source blocks don't have copy to clipboard functionality working in them. I'll Look into it.

Vandit1604 commented:
All the pages that had overlapping TOC are fixed now.

I will write this here so i don't forget later.
If someone is writing source block on the start of the page.

`[.w-60]` add this to change the width of the source block.

```
[.w-60]
[source,bash]
----
# pkg install jenkins-lts
----
```

Vandit1604 commented:
Its working :tada: 

- [x] Developer Docs has copy to clipboard functionality.
- [x] User Handbook has copy to clipboard functionality. 
- [x]  Table of Content is not overlapping any source block.

I think its ready for another thorough review.

krisstern commented:
Thanks for all the hard work @Vandit1604! We will review this PR carefully shortly.

zbynek reviewed (changes requested):
The idea with 60% width is not working very well for smaller screens (and will break if content changes). Also adding empty paragraph above to avoid conflict with TOC doesn't feel right and only works for some screen sizes. Instead
* by default code blocks should have `clear:both` so that they are below any floating elements
* where this would create too much space because the TOC is too long, the script blocks can use something like
```
   display: flex;
   min-width: fit-content;
   flex-direction: column;
```

zbynek reviewed (approved):
I didn't check all the affected pages, but looks good in general :tada:

kmartens27 reviewed (commented):
These look great, thanks very much @Vandit1604.

For the reverse proxy pages, I found that the `context-path.adoc` file has the copy to clipboard button when included in the troubleshooting page, but not when it is included in the various configuration pages. The only difference I could find between the two is that the configuration pages are set as `subsection` as opposed to `section`, but I don't know why that would prevent the copy button from showing up.
![Screenshot 2023-04-14 at 4 47 34 PM](https://user-images.githubusercontent.com/99040580/232153609-dbaf5218-cb4e-4321-98eb-713de044f523.png)

Vandit1604 commented:
> These look great, thanks very much @Vandit1604.
> 
> For the reverse proxy pages, I found that the `context-path.adoc` file has the copy to clipboard button when included in the troubleshooting page, but not when it is included in the various configuration pages. The only difference I could find between the two is that the configuration pages are set as `subsection` as opposed to `section`, but I don't know why that would prevent the copy button from showing up. ![Screenshot 2023-04-14 at 4 47 34 PM](https://user-images.githubusercontent.com/99040580/232153609-dbaf5218-cb4e-4321-98eb-713de044f523.png)

Thanks @kmartens27 for your review.
If that's the case would changing the `layout: subsection` to `layout: section` permissible?
I would check locally if changing `layout: value` would solve the issue completely.

krisstern commented:
I have just modified the `content/_layouts/subsection.html.haml` file to add the `.ctc` class in so that the copy-to-cliipboard button would show.

krisstern commented:
Have just checked all Developer Guide and User Guide page code snippets.

krisstern reviewed (approved):
Looks good!

Vandit1604 commented:
If there is any more work to be done on the PR, please let me know.

kmartens27 reviewed (approved):
Looks great, thanks very much @Vandit1604!

I think we should wait for a security review before merging, just to be sure that nothing is broken/regressed unexpectedly.

kmartens27 commented:
@jenkins-infra/security would someone be able to review this PR and confirm/deny whether there are any issues with adding this functionality? Thanks so much!

kmartens27 commented:
discussed with @MarkEWaite, merging as there do not appear to be any issues

Vandit1604 commented:
Thank You! @krisstern for the help in this pull request.

krisstern commented:
You are welcome @Vandit1604
